src/send-message.ts

```diff
@@ -0,0 +1,176 @@
+}
+
+function parseStreamDestination(
+  input: string
+): StreamDestination["destination"] {
+  return allNumeric(input) ? Number.parseInt(input) : input;
+}
+
+function parseDestinationDetails({
+  kind,
+  destination,
+}: {
+  kind: DestinationKind;
+  destination: string;
+}): Result<DestinationDetails, string> {
+  switch (kind) {
+    case DestinationKind.Private: {
+      return Ok({
+        kind,
+        destination: parsePrivateMessageDestinations(destination),
+      });
+    }
+
+    case DestinationKind.Stream: {
+      return getMandatoryInputFromJob("topic")
+        .map((topic) => {
+          return {
+            kind,
+            topic,
+            destination: parseStreamDestination(destination),
+          };
+        })
+        .mapErr(() => 'topic is mandatory when type is "stream"');
+    }
+
+    default: {
+      assertExhaustiveSwitch(kind);
+    }
+  }
+}
+
+function getMandatoryInputFromJob(key: string): Result<string, string> {
+  const candidate = getInput(key, { required: true });
+
+  return candidate
+    ? Ok(candidate)
+    : Err(`input "${key}" must be provided and non-empty`);
+}
+
+// TypeScript lacks a way to genericize this cleanly to
+// getEnumInputFromGHA<T extends Enum>(key: string, spec: T),
+// see https://github.com/microsoft/TypeScript/issues/30611
+function getDestinationKindInputFromJob(): Result<DestinationKind, string> {
+  const validValues = Object.values(DestinationKind);
+  const errorMessage = `input "type" must exist and be one of ${validValues.join(
+    " ,"
+  )}`;
+  return getMandatoryInputFromJob("type")
+    .mapErr(() => errorMessage)
+    .andThen((input) => {
+      return validValues.includes(input as any)
+        ? Ok(input as DestinationKind)
+        : Err(errorMessage);
+    });
+}
+
+function getDestinationDetails(): Result<DestinationDetails, string> {
+  return getDestinationKindInputFromJob()
+    .andThen((destinationKind) => {
+      return getMandatoryInputFromJob("to").map((destination) => {
+        return { kind: destinationKind, destination };
+      });
+    })
+    .andThen(parseDestinationDetails);
+}
+
+async function getZulipClient(): Promise<Result<Client, string>> {
+  return new AsyncResultWrapper(
+    Result.all(
+      getMandatoryInputFromJob("api-key"),
+      getMandatoryInputFromJob("email"),
+      getMandatoryInputFromJob("organization-url")
+    )
+  )
+    .flatMap(
+      ([apiKey, username, realm]) =>
+        new AsyncResultWrapper(
+          Result.wrapAsync(async () =>
+            zulipInit({
+              apiKey,
+              username,
+              realm,
+            })
+          )
+        )
+    )
+    .resolve();
+}
+
+async function postMessage(): Promise<Result<string, string>> {
+  const topic = getInput("topic", { required: false });
+  return new AsyncResultWrapper(
+    Result.all(
+      getMandatoryInputFromJob("content"),
+      getDestinationDetails(),
+      await getZulipClient()
+    )
```

Comment:
It may make sense to instead use [`AsyncResult.all`](https://github.com/movesthatmatter/ts-async-results#combining-results), though the ergonomics are a little weirder there. It would probably replace the manual call to `new AsyncResultWrapper` at the top level, though may create two new calls to it on L144+L145. I'll do some investigation on this Monday.

Reply:
I assume that using `[AsyncResult.all](https://github.com/movesthatmatter/ts-async-results#combining-results)` didn't bear any fruit?